server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs

```diff
@@ -348,46 +340,53 @@ produceBatch rtqOperationType rtqRemoteSchemaInfo RemoteRelField{..} batchInputs
         fcName
         (rtrRemoteFields
            (rmfRemoteRelationship rrRemoteField))
-    batchRemoteTopQuery = VQ.RemoteTopField{..}
-    rtqFields =
-      flip map indexedRows $ \(i, variables) ->
-         fieldCallsToField
-           rrArguments
-           variables
-           rrSelSet
-           (Just (arrayIndexAlias i))
-           (rtrRemoteFields remoteRelationship)
-    indexedRows = zip (map ArrayIndex [0 :: Int ..]) $ toList $ biRows batchInputs
-    batchIndices = map fst indexedRows
-    remoteRelationship = rmfRemoteRelationship rrRemoteField
     -- TODO These might be good candidates for reusing names with DuplicateRecordFields if they have the same semantics:
     batchRelationshipKeyIndex = rrAliasIndex
     batchPhantoms = rrPhantomFields
     batchRelFieldPath = rrRelFieldPath
 
--- | Produce the alias name for a result index.
-arrayIndexAlias :: ArrayIndex -> G.Alias
-arrayIndexAlias i =
-  G.Alias (G.Name (arrayIndexText i))
+-- | Tag 'biRows' list with enumerated aliases we can use to recover the
+-- original order of the input list here, from the remote results (even if the
+-- remote is not well-behaved wrt ordering.
+--
+-- This must satisfy (modulo newtypes)
+--
+-- @
+--   id =  unEnumerateRowAliases . enumerateRowAliases
+-- @
+enumerateRowAliases :: [row] -> [(G.Alias, row)]
+enumerateRowAliases = map (over _1 alias) . zip [0::Int ..] where
+  alias i = G.Alias $ G.Name $ "hasura_array_idx_" <> T.pack (show i)
+
+-- | re-sort and return aliased results, following 'enumerateRowAliases'
+--
+-- 'Nothing' here means either a bug or a very naughty remote.
+unEnumerateRowAliases :: [(Text, row)] -> Maybe [row]
+unEnumerateRowAliases = fmap (map snd . sortOn fst) . mapM (mapMOf _1 toInt)
+  where
+    -- NOTE: a naive lexicographic sort here will be incorrect when len > 9
+    toInt s = do 
+      ns <- T.stripPrefix hasuraAliasTag s
+      readMaybe (T.unpack ns) :: Maybe Int
 
--- | Produce the alias name for a result index.
-arrayIndexText :: ArrayIndex -> Text
-arrayIndexText (ArrayIndex i) =
-  T.pack ("hasura_array_idx_" ++ show i)
+hasuraAliasTag :: Text
+hasuraAliasTag = "hasura_array_idx_"
 
 -- | Produce a field from the nested field calls.
 fieldCallsToField ::
      Map.HashMap G.Name AnnInpVal
   -> Map.HashMap G.Variable G.ValueConst
   -> SelSet
-  -> Maybe G.Alias
+  -> G.Alias
   -> NonEmpty FieldCall
   -> Field
-fieldCallsToField userProvidedArguments variables finalSelSet = nest
+fieldCallsToField userProvidedArguments variables finalSelSet topAlias = 
```

Comment:
is "Call" a standard terminology? `FieldCall` seems like a bad name if it isn't derivable from standard terminology no?

Reply:
Yeah I did think about this a little, but then gave up. I think we can come up with a better name for `fieldCallsToField` for sure. 

`FieldCall` might be a good enough name with good docs. It's a "field" without a selection set, where the spec says "Object fields are conceptually functions which yield values." (which is already confusing terminology, but should be familiar). But no "field call" isn't standard terminology afaict